### ISSUE TITLE
Vulnerbility report - Established WebSocket session remains open

### ISSUE THREAD

**Issue by dz2742-sponsor** [matusfaro/quarantab#2]
Hi,
I came for the bug bounty and I found one, what is the recommended way of communicating the details?
Thank you

**Comment by matusfaro** [comment]
Hi @dz2742,

Honestly I just released this extension yesterday so a zero-day is probably not an issue to discuss publicly. You can post it here.

**Comment by dz2742-sponsor** [comment]
Changing proxy settings doesn't disconnect established WebSocket session
PoC:
```HTML
<!DOCTYPE html>
<html>
<script>
const websocket = new WebSocket("wss://" + location.host + "/ws");
let logs = "";
websocket.addEventListener('message', event => {
  logs = event.data + "\n" + logs;
  document.body.innerText = logs;
});
setInterval(()=>{
  websocket.send(Math.random());
}, 1000);
</script>
<img src="/delay">
</html>
```

**Comment by matusfaro** [comment]
Nice catch, thank you for your work!

I think that constitutes a full exploit and deserves the entire 100 USD in the pool. Would you like to donate some of that back for future reports? Let me know if so.

Otherwise, do you accept bitcoin/cardano? Post your address here or send me an email at matus at matus io

**Comment by dz2742-sponsor** [comment]
My pleasure, also congratulations on trending on HN front page!

I would like to donate to other projects on Github, would you mind making a few one-time sponsorships in my place?

**Comment by matusfaro** [comment]
Yes of course and thank you!

Do you have specific projects in mind? If not, I will look for some that I've used recently and post back here.

**Comment by dz2742-sponsor** [comment]
- https://github.com/KDE
- https://github.com/ajayyy-org
- https://github.com/TeamPiped
- https://github.com/organicmaps

I would like to donate 20 USD to each of them and the remaining 20 USD is yours, thank you :smiley:

**Comment by matusfaro** [comment]
Sponsorships [sent](https://github.com/matusfaro?tab=sponsoring)

I'll think of you dz2742 when I buy a beer with that twenty. :)

P.S. Feel free to submit more vulnerabilities, I replenished the pool for another 100 USD.

**Comment by matusfaro** [comment]
As far as my research goes, there doesn't seem to be a way to cut network access for in-progress connections from an Addon.

[Suppression for websockets](https://github.com/matusfaro/quarantab/commit/adb692d3a48aa294092aea5059f56cfc9d8ae38b) specifically is to simply block all websockets in our container even before network lock is requested.

However I believe the vulnerability still exists for other connections so I am leaving this issue open while I look into it.

Particularly, any request body can be a [ReadableStream](https://developer.mozilla.org/en-US/docs/Web/API/Request/body) that I predict will stay open for as long as I keep the stream open. I will do a POC to verify this.

Any suggestions on how to fix this are welcome.

**Comment by Alan-Liang** [comment]
It seems that WebRTC also enables data leaks, which is totally not surprising.

TL; DR: Established WebRTC session remains open.

A simple POC using peerjs:

1. Open <https://lab.olic.link/rtc/attacker.html> in a normal browser tab
2. Open <https://lab.olic.link/rtc/victim.html> in an isolated tab
3. Wait for victim.html to finish connection
4. Verify that the WS connection on victim.html has been closed (the WS is for connection brokering only, it is possible to leak data without a WS connection; it's there just because peerjs use websockets.)
5. Click "block network"
6. Verify that message could still be sent from victim.html to attacker.html

**Comment by dz2742-sponsor** [comment]
Firefox has an offline mode (Alt key to bring out the menu bar - File - Work Offline) from the good old days, enabling it cuts existing connections including WebSocket. It's browser-wide so other tabs (especially video conference) will be affected.
https://addons.mozilla.org/en-US/firefox/addon/work-offline-page-only/
https://github.com/Emano-Waldeck/work-offline/

**Comment by matusfaro** [comment]
@dz2742

> Firefox has an offline mode

Unfortunately, it's not possible to control this through an extension. The [homepage](https://add0n.com/work-offline.html) of the first extension contains "...except already connected WebSocket connections" and the second link [uses](https://github.com/Emano-Waldeck/work-offline/blob/master/v2/background.js#L89) `chrome.webRequest.onBeforeRequest` which only intercepts new requests as well.

However, there are three attack vectors I identified and here are the best solutions I could come up with:

1. Websockets: Simply block all websockets for the entire duration of the Container even before "block network".
2. WebRTC: Disable WebRTC globally in Firefox for the duration of our temporary Containers. (Not possible to enable per tab/container)
3. Long streaming/polling HTTP requests: Add a listener for request open and close and keep a running count of active connections. Wait until all connections are closed before indicating to the user the network is locked.

Interesting note:
- Browser WebRTC implementations typically use websockets for signalling including your POC. Simply blocking websockets broke WebRTC, but in theory you could do signalling over HTTP long polling as well so it's not a reliable way to block WebRTC.

Other things I considered:
- Reopen/refresh a tab with network lock, hoping all assets are cached. (Still wouldn't block webrtc, but could it work without active signalling?)
- Inject scripts into the site to attempt to intercept/close webrtc/websockets/http requests (Too fragile)

**Comment by dz2742-sponsor** [comment]
> Unfortunately, it's not possible to control this through an extension. The [homepage](https://add0n.com/work-offline.html) of the first extension contains "...except already connected WebSocket connections" and the second link [uses](https://github.com/Emano-Waldeck/work-offline/blob/master/v2/background.js#L89) chrome.webRequest.onBeforeRequest which only intercepts new requests as well.

According to my testing the extension does cut out existing WebSocket, the second link is its source code.

**Comment by matusfaro** [comment]
> According to my testing the extension does cut out existing WebSocket, the second link is its source code.

Thank you for this insight.

I found that the extension does also [inject a content script into the page](https://github.com/Emano-Waldeck/work-offline/blob/master/v2/data/inject/offline.js#L17) and calls `window.stop()` which does indeed kill any existing WebSocket connections. This means I can both prevent new connections as well as stop live connections for WebSockets.

For WebRTC however `window.stop()` has no effect based on my testing. And even if it did, I have no way of preventing new WebRTC connections to start. So I will still have to disable WebRTC globally for the browser for the entire session of my temporary container.

**Comment by matusfaro** [comment]
This is now fixed in released version 0.0.3.

Thanks for all your help!
